Source/ACE/Entity/Player.cs

```diff
@@ -133,6 +130,67 @@ private Position PositionLastPortal
             }
         }
 
+        public bool UnknownSpell(uint spellId)
+        {
+            return !(AceObject.SpellIdProperties.Exists(x => x.SpellId == spellId));
+        }
+
+        public void HandleMagicRemoveSpellId(uint spellId)
+        {
+            if (!AceObject.SpellIdProperties.Exists(x => x.SpellId == spellId))
+            {
+                log.Error("Invalid spellId passed to Player.RemoveSpellFromSpellBook");
+                return;
+            }
+            ActionChain unlearnSpellChain = new ActionChain();
+            unlearnSpellChain.AddAction(
+               this,
+               () =>
+                   {
+                       AceObject.SpellIdProperties.RemoveAt(AceObject.SpellIdProperties.FindIndex(x => x.SpellId == spellId));
+                       GameEventMagicRemoveSpellId removeSpellEvent = new GameEventMagicRemoveSpellId(Session, spellId);
+                       Session.Network.EnqueueSend(removeSpellEvent);
+                   });
+            unlearnSpellChain.EnqueueChain();
+        }
+
+        public void HandleLearnSpell(uint spellId)
+        {
+            ActionChain learnSpellChain = new ActionChain();
+            SpellTable spells = SpellTable.ReadFromDat();
+            if (!spells.Spells.ContainsKey(spellId))
```

Comment:
Not sure if that is cached or not, but we probably want to load that table in memory at startup I would think.   It is all that big.